src/arcsf/data/data_module.py

```diff
@@ -107,34 +132,144 @@ def __init__(
+            qa[0], return_tensors="pt", add_special_tokens=True
+        )
+        encoded_tar = self.tokenizer(
+            qa[1], return_tensors="pt", add_special_tokens=True
+        )
+        return (encoded_inp, encoded_tar)
+
+    def batch_formatter(
+        self,
+        qa: tuple[str, str, int],
+        padding_flag: int = 1,
+    ) -> dict[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Formats the question-answer pair with padding and appropriately masked labels
+            allowing for batch computation.
+
+        Args:
+            qa : Tuple containing a question--answer pair
+
+        Returns:
+            formatted : formatted version of the input which can readily be passed to
+            model
+        """
+        question, _ = qa
+        encoded = self.tokenizer(
+            self.qa_formatter(qa),
+            max_length=self.max_length,
+            truncation=False,
+        )
+        num_question_tokens = len(self.tokenizer.tokenize(question))
+        pad_length = self.max_length - len(encoded.input_ids)
+        padded = (
+            encoded["input_ids"]
+            + padding_flag * [self.tokenizer.eos_token_id] * pad_length
+        )
+        pad_attention_mask = encoded["attention_mask"] + padding_flag * [0] * pad_length
+        label = encoded.input_ids
+        if len(encoded.input_ids) == self.max_length:
+            label = encoded.input_ids
+        else:
+            label = (
+                encoded["input_ids"]
+                + padding_flag * [self.tokenizer.eos_token_id]
+                + padding_flag * [-100] * (pad_length - 1)
+            )
+
+        # change label to -100 for question tokens
+        for i in range(num_question_tokens):
+            label[i] = -100
+        return {
+            "input_ids": torch.tensor(padded).to(self.device),
+            "labels": torch.tensor(label).to(self.device),
+            "attention_mask": torch.tensor(pad_attention_mask).to(self.device),
+        }
+
+    def get_perturbed(
+        self, qa: tuple[str, str], row: int
+    ) -> list[dict[torch.Tensor, torch.Tensor, torch.Tensor]]:
+        """Returns batch-formatted version of the question answer pair along with
+            perturbed, erroneous question-answer pairs. For evaluation purposes.
+
+        Args:
+            qa : ground truth question answer tuple
+            row : question row ID for obtaining perturbed input
+
+        Returns:
+            formatted: batch formatted version of the perturbed inputs along with
+                formatted ground truth inputs
+        """
+        inp, _ = qa
+        author_n = self.data[row]["author_index"]
+        question_n = self.data[row]["question_index"]
+
+        with hf_progress_bars_disabled():
+            perturbed_options = self.data.filter(
+                lambda sample: sample["author_index"] == author_n
+                and sample["question_index"] != question_n
+            ).shuffle(seed=self.rand_gen.seed())
```

Comment:
It will give the same question indices for each question. We can remove the seed if this isn't the functionality we want?

Reply:
@J-Dymond - I'm still unsure about this, but in two ways:

- Looking at the docs of [torch.Generator.seed](https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator.seed) this will not produce a reproducible value (i.e. the fact the generator is initialised with a seed of 42 has no bearing on what `rand_gen.seed()` will return). I think the correct thing to do would be to use the [`generator` arg of `shuffle`](https://huggingface.co/docs/datasets/v2.20.0/en/package_reference/main_classes#datasets.Dataset.shuffle), but note this is only compatible with `numpy` generators not `torch` generators.

- I'm not sure we want the same perturbed answers to be returned every time this is called with the same `author_n` and `question_n` (within a script), but I'm not sure this is a big issue either.

Some of this may not matter once we have a dataset with actual perturbed answers.

-

Reply:
Ok, I misunderstood this, sorry! I've added another attribute storing the random_seed. But I agree, we don't need this with the new dataset, and I'm not sure we want it either.